lib/charms/operator_libs_linux/v0/grub.py

```diff
@@ -29,4 +37,259 @@
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> str:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def _data(self) -> Dict[str, str]:
+        """Data property."""
+        data = self._lazy_data
+        if data is None:
+            data = self._lazy_data = _load_config(GRUB_CONFIG)
+
+        return data
+
+    def _apply(self):
+        """Check if an update to /boot/grub/grub.cfg is available."""
+        try:
+            subprocess.check_call(f"grub-mkconfig -o {GRUB_CONFIG}", stderr=subprocess.STDOUT)
+            subprocess.check_call(["/usr/sbin/update-grub"], stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as error:
+            logger.exception(error)
+            raise ApplyError("New config check failed.") from error
+
+    def _save_grub_configuration(self) -> None:
+        """Save current gru configuration."""
+        logger.info("saving new grub config to %s", GRUB_CONFIG)
+        applied_configs = {self.path, *self.applied_configs}  # using set to drop duplicity
+        registered_configs = os.linesep.join(
+            FILE_LINE_IN_DESCRIPTION.format(path=path) for path in applied_configs
+        )
+        header = CONFIG_HEADER + CONFIG_DESCRIPTION.format(configs=registered_configs)
+        _save_config(GRUB_CONFIG, self._data, header)
+
+    def _set_value(self, key: str, value: str) -> bool:
+        """Set new value for key."""
+        logger.debug("setting new value %s for key %s", value, key)
+        current_value = self._data.get(key)
+        if current_value is None:
+            self._data[key] = value
+            return True
+
+        if current_value != value:
+            logger.warning(
+                "tries to overwrite key %s, which has value %s, with value %s",
+                key,
+                current_value,
+                value,
+            )
+            raise ValidationError(
+                f"key {key} already exists and its value is {current_value}", key
+            )
+
+        return False
+
+    def _update(self, config: Dict[str, str]) -> Set[str]:
+        """Update data in object."""
+        logger.debug("updating current config")
+        changed_keys = set()
+        for key, value in config.items():
+            changed = self._set_value(key, value)
+            if changed:
+                changed_keys.add(key)
+
+        return changed_keys
+
+    @property
+    def charm_name(self) -> str:
+        """Get charm name or use value obtained from JUJU_UNIT_NAME env."""
+        if self._charm_name is None:
+            self._charm_name, *_ = os.getenv("JUJU_UNIT_NAME", "unknown").split("/")
```

Comment:
To me it seems odd that this is so closely tied to the charm name, particularly the use of environment variables. What's the need for having the charm name in this class? And if needed, I think it'd be preferable to require it to be passed in, rather than parsing environment variables in a library.

Reply:
To be able to store only charm config file at `/etc/default/grub.d/90-juju-<charm-name>`. Using the env variable is only there so the charm developer does not have to think about it, we do not want to use `ops` here. Do you think that it's not needed? I can remove it.

Reply:
Yeah, I would prefer charm libs don't parse environment variables directly, but that the charm name is passed in explicitly to the initialiser.

Reply:
Ok, I'm going to remove it.